crates/runtime/src/instance/allocator/default.rs

```diff
@@ -0,0 +1,292 @@
+            }
+
+            unsafe {
+                let mem_slice = Self::get_memory_slice(init, instance);
+                let end = start + init.data.len();
+                let to_init = &mut mem_slice[start..end];
+                to_init.copy_from_slice(&init.data);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+unsafe impl InstanceAllocator for DefaultInstanceAllocator {
+    unsafe fn allocate(
+        &self,
+        mut req: InstanceAllocationRequest,
+    ) -> Result<InstanceHandle, InstantiationError> {
+        debug_assert!(!req.externref_activations_table.is_null());
+        debug_assert!(!req.stack_map_registry.is_null());
+
+        let memories = self.create_memories(&req.module)?;
+        let tables = Self::create_tables(&req.module);
+
+        let handle = {
+            let instance = Instance {
+                allocator: self,
+                module: req.module.clone(),
+                offsets: VMOffsets::new(std::mem::size_of::<*const u8>() as u8, &req.module),
+                memories,
+                tables,
+                dropped_elements: RefCell::new(EntitySet::with_capacity(
+                    req.module.passive_elements.len(),
+                )),
+                dropped_data: RefCell::new(EntitySet::with_capacity(req.module.passive_data.len())),
+                host_state: req.host_state.take(),
+                vmctx: VMContext {},
+            };
+            let layout = instance.alloc_layout();
+            let instance_ptr = alloc::alloc(layout) as *mut Instance;
+            if instance_ptr.is_null() {
+                alloc::handle_alloc_error(layout);
+            }
+            ptr::write(instance_ptr, instance);
+            InstanceHandle::new(instance_ptr)
+        };
+
+        let instance = handle.instance();
+        initialize_vmcontext(
+            instance,
+            &req,
+            &|index| instance.memories[index].vmmemory(),
+            &|index| instance.tables[index].vmtable(),
+        );
+
+        Ok(handle)
+    }
+
+    unsafe fn initialize(
+        &self,
+        handle: &InstanceHandle,
+        is_bulk_memory: bool,
+        data_initializers: Arc<[OwnedDataInitializer]>,
+    ) -> Result<(), InstantiationError> {
+        // Check initializer bounds before initializing anything. Only do this
+        // when bulk memory is disabled, since the bulk memory proposal changes
+        // instantiation such that the intermediate results of failed
+        // initializations are visible.
+        if !is_bulk_memory {
+            Self::check_table_init_bounds(handle.instance())?;
+            Self::check_memory_init_bounds(handle.instance(), data_initializers.as_ref())?;
+        }
+
+        // Apply fallible initializers. Note that this can "leak" state even if
+        // it fails.
+        Self::initialize_tables(handle.instance())?;
+        Self::initialize_memories(handle.instance(), data_initializers.as_ref())?;
+
+        Ok(())
+    }
```

Comment:
It isn't clear to me that implementations will ever do this initialization step differently. Do you expect it to change with instance pooling? If not, then I think we don't need to make `initialize` part of the trait.

Reply:
I do expect there to be some differences with the allocator implementations with respect to table / linear memory initialization.  `userfaultfd` support can lazily initialize the pages, so it might simply take a reference on the initializers for later use, for example.